File: atak/ATAKMapEngine/lib/src/nonpri-common/java/com/atakmap/map/layer/raster/mosaic/MosaicUtils.java

package com.atakmap.map.layer.raster.mosaic;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.util.HashSet;
import com.atakmap.coremap.locale.LocaleUtil;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

import org.gdal.gdal.Dataset;
import org.gdal.gdal.gdal;

import com.atakmap.coremap.log.Log;
import com.atakmap.coremap.maps.coords.GeoPoint;

import com.atakmap.io.ZipVirtualFile;
import com.atakmap.map.layer.raster.DatasetDescriptor;
import com.atakmap.map.layer.raster.DatasetProjection2;
import com.atakmap.map.layer.raster.ImageDatasetDescriptor;
import com.atakmap.map.layer.raster.gdal.GdalDatasetProjection2;
import com.atakmap.map.layer.raster.gdal.GdalLayerInfo;
import com.atakmap.map.layer.raster.pfps.PfpsUtils;
import com.atakmap.math.NoninvertibleTransformException;
import com.atakmap.math.PointD;
import com.atakmap.math.Matrix;
import com.atakmap.android.maps.tilesets.TilesetInfo;

public class MosaicUtils {

    public final static String TAG = "MosaicUtils";

    public final static String SUBDIR_NITF = "nitf";
    public final static String SUBDIR_GEOTIFF = "geotiff";
    public final static String SUBDIR_MRSID = "mrsid";
    public final static String SUBDIR_RPF = "rpf";



    private final static int INTERVAL = 50;

    private final static Set<String> SUBDIRS = new HashSet<String>();
    static {
        SUBDIRS.add(SUBDIR_NITF);
        SUBDIRS.add(SUBDIR_GEOTIFF);
        SUBDIRS.add(SUBDIR_MRSID);
    }

    public final static int DEFAULT_NUM_LEVELS = 3;

    private MosaicUtils() {
    }

    public interface BuildMosaicCallback{
        public void onProgressUpdate(int itemsProcessed);
    }

    public static boolean isMosaicDir(File f) {
        return isMosaicDir(f, Integer.MAX_VALUE);
    }
    
    public static boolean isMosaicDir(File f, int limit){
        if (!f.isDirectory())
            return false;
        if (PfpsUtils.isPfpsDataDir(f))
            return true;
        // didn't look like PFPS, check for our directories
        int hits = 0;
        File[] children = f.listFiles();
        if (children != null) { 
            int checkLimit = Math.min(children.length, limit);
            for (int i = 0; i < checkLimit; i++) {
                if (SUBDIRS.contains(children[i].getName().toLowerCase(LocaleUtil.getCurrent()))) {
                    if (!children[i].isDirectory())
                        continue;

                    File[] files = children[i].listFiles();
                    if (files != null && files.length > 0)
                        hits++;
                }
            }
        }

        return (hits > 0);
    }

    public static int buildMosaicDatabase(File mosaicDir, File databaseFile) {
        return buildMosaicDatabase(mosaicDir, databaseFile, null);
    }

    public static int buildMosaicDatabase(File mosaicDir, File databaseFile, BuildMosaicCallback callback) {
        AtomicInteger count = new AtomicInteger(0);

        MosaicDatabaseBuilder2 database = null;
        try {
            database = ATAKMosaicDatabase3.create(databaseFile);

            URI relativeTo = mosaicDir.toURI();
    
            
    
            database.beginTransaction();
            try {
                File[] subdirs = mosaicDir.listFiles();
                if (subdirs != null) { 
                    for (int i = 0; i < subdirs.length; i++) {
                        if (!subdirs[i].isDirectory())
                            continue;
                        if (subdirs[i].getName().equalsIgnoreCase(SUBDIR_RPF)) {
                            PfpsUtils.createRpfDataDatabase2(database, subdirs[i]);
                        } else if (SUBDIRS.contains(subdirs[i].getName().toLowerCase(LocaleUtil.getCurrent()))) {
                            buildGenericSubDatabase(database, subdirs[i], relativeTo, count, callback);
                        }
                    }
                }

                if(callback != null)
                    callback.onProgressUpdate(-100);

                database.createIndices();
                database.setTransactionSuccessful();
            } finally {
                database.endTransaction();                
            }
        } finally {
            if(database != null)
                database.close();
        }
        
        return count.get();
    }

    private static void buildGenericSubDatabase(MosaicDatabaseBuilder2 database, File subdir,
            URI relativeTo, AtomicInteger count, BuildMosaicCallback callback) {
        
        final String genericType = subdir.getName().toLowerCase(LocaleUtil.getCurrent());
        final boolean nitfChecks = genericType.equals("nitf");
        File[] children = subdir.listFiles();
        if (children == null) 
            children = new File[0];

       
        Dataset dataset;
        DatasetProjection2 proj;
        int width;
        int height;
        PointD scratch = new PointD(0, 0);
        GeoPoint ul = GeoPoint.createMutable();
        GeoPoint ur = GeoPoint.createMutable();
        GeoPoint lr = GeoPoint.createMutable();
        GeoPoint ll = GeoPoint.createMutable();
        String type;
        boolean isDefined;
        String path;
        int srid;
        boolean isPrecisionImagery;

        for (int i = 0; i < children.length; i++) {
            int tempCount = count.getAndIncrement();
            if(callback != null && (tempCount % INTERVAL) == 0){
                callback.onProgressUpdate(tempCount);
            }

            if (children[i].isDirectory()) {
                Log.w(TAG, "Skipping " + subdir.getName() + " subdirectory, " + children[i]);
                continue;
            }

            type = null;
            width = 0;
            height = 0;
            isDefined = false;
            isPrecisionImagery = false;
            srid = -1;

            path = children[i].getAbsolutePath();
            if(nitfChecks) {
            }

            if(!isDefined) {
                dataset = null;
                try {
                    dataset = gdal.Open(path);
                    if (dataset == null) {
                        Log.w(TAG, "Unable to open mosaic frame: " + children[i]);
                        continue;
                    }

                    proj = GdalDatasetProjection2.getInstance(dataset);

                    if (proj == null) {
                        Log.w(TAG, "Unable to create dataset projection: " + children[i]);
                        continue;
                    }

                    width = dataset.GetRasterXSize();
                    height = dataset.GetRasterYSize();

                    scratch.x = 0;
                    scratch.y = 0;
                    proj.imageToGround(scratch, ul);
                    scratch.x = width;
                    scratch.y = 0;
                    proj.imageToGround(scratch, ur);
                    scratch.x = width;
                    scratch.y = height;
                    proj.imageToGround(scratch, lr);
                    scratch.x = 0;
                    scratch.y = height;
                    proj.imageToGround(scratch, ll);
                
                    srid = ((GdalDatasetProjection2)proj).getNativeSpatialReferenceID();

                    type = genericType;
                } finally {
                    if (dataset != null)
                        dataset.delete();
                }
            }
            final double gsd = DatasetDescriptor.computeGSD(width, height, ul, ur, lr, ll);
            database.insertRow(relativeTo.relativize(children[i].toURI()).getPath(),
                               type,
                               isPrecisionImagery,
                               ul,
                               ur,
                               lr,
                               ll,
                               gsd * (double)(1<<DEFAULT_NUM_LEVELS),
                               gsd,
                               width,
                               height,
                               srid);
            }
        }
}

File: takengine/mapengine/sdk/src/raster/mosaic/MosaicUtils.cpp
#include "MosaicUtils.h"
#include "raster/PrecisionImageryFactory.h"
#include "raster/pfps/PfpsUtils.h"
#include "raster/gdal/GdalDatasetProjection.h"
#include "raster/gdal/GdalLayerInfo.h"
#include "raster/DatasetDescriptor.h"
#include "raster/ImageDatasetDescriptor.h"
#include "raster/PrecisionImageryFactory.h"
#include "util/IO.h"
#include "math/Utils.h"

using namespace TAK::Engine::Raster;
using namespace TAK::Engine::Util;

namespace atakmap {
    namespace raster {
        namespace mosaic {

            namespace {
                const int INTERVAL = 50;
                const char *SUBDIR_NITF = "nitf";
                const char *SUBDIR_PRI = "pri";
                const char *SUBDIR_PFI = "pfi";
                const char *SUBDIR_GEOTIFF = "geotiff";
                const char *SUBDIR_MRSID = "mrsid";
                const char *SUBDIR_RPF = "rpf";

                const char *SUBDIRS[] = {
                    SUBDIR_NITF,
                    SUBDIR_PRI,
                    SUBDIR_PFI,
                    SUBDIR_GEOTIFF,
                    SUBDIR_MRSID,
                    nullptr
                };

                bool isInSubdirs(const char *s) {
                    int i = 0;
                    while (SUBDIRS[i]) {
                        if (strcmp(SUBDIRS[i], s) == 0)
                            return true;
                        i++;
                    }
                    return false;
                }

            }


            bool MosaicUtils::isMosaicDir(const char *f, int limit)
            {
                if (!util::isDirectory(f))
                    return false;
                if (pfps::PfpsUtils::isPfpsDataDir(f))
                    return true;
                // didn't look like PFPS, check for our directories
                int hits = 0;
                std::vector<std::string> children = util::getDirContents(f);

                int checkLimit = math::min((int)children.size(), limit);
                for (int i = 0; i < checkLimit; i++) {
                    std::string clower = util::toLowerCase(util::getFileName(children[i].c_str()));
                    if (isInSubdirs(clower.c_str())) {
                        if (!util::isDirectory(children[i].c_str()))
                            continue;
                        if (util::getDirContents(children[i].c_str()).size() > 0)
                            hits++;
                    }
                }

                return (hits > 0);

            }

            void MosaicUtils::buildMosaicDatabase(const char *mosaicDir, const char *databaseFile,
                                                  BuildMosaicCallback *callback)
            {
                ATAKMosaicDatabase database;

                int count = 0;

                try {
                    database.open(databaseFile);
                    database.beginTransaction();

                    std::vector<std::string> subdirs = util::getDirContents(mosaicDir);
                    for (size_t i = 0; i < subdirs.size(); i++) {
                        if (!util::isDirectory(subdirs[i].c_str()))
                            continue;
                        std::string lowerFname = util::toLowerCase(util::getFileName(subdirs[i].c_str()));
                        const char *clowerFname = lowerFname.c_str();
                        if (strcmp(SUBDIR_RPF, clowerFname) == 0) {
                            pfps::PfpsUtils::createRpfDataDatabase(&database, subdirs[i].c_str());
                        } else if (strcmp(SUBDIR_MRSID, clowerFname) == 0) {
#if 0
                            // mr. sid disabled and fallback to generic for now per Chris
                            buildMrsidSubDatabase(&database, subdirs[i].c_str(), mosaicDir, &count, callback);
#else
                            buildGenericSubDatabase(&database, subdirs[i].c_str(), mosaicDir, &count, callback);
#endif
                        } else if (isInSubdirs(clowerFname)) {
                            buildGenericSubDatabase(&database, subdirs[i].c_str(), mosaicDir, &count, callback);
                        }
                    }

                    database.setTransactionSuccessful();
                    database.endTransaction();
                    database.close();
                } catch (...) {
                    database.endTransaction();
                    database.close();
                    throw;
                }
            }

            void MosaicUtils::buildGenericSubDatabase(ATAKMosaicDatabase *database, const char *subdir,
                                                      const char *relativeTo, int *count,
                                                      BuildMosaicCallback *callback)
            {
                std::string genericType = util::toLowerCase(util::getFileName(subdir));
                bool nitfChecks = genericType.compare("nitf") == 0;
                std::vector<std::string> children = util::getDirContents(subdir);
                int width;
                int height;
                math::PointD scratch(0, 0);
                core::GeoPoint ul;
                core::GeoPoint ur;
                core::GeoPoint lr;
                core::GeoPoint ll;
                std::string type;
                bool isDefined;
                //String path;
                for (size_t i = 0; i < children.size(); i++) {
                    int tempCount = *count;
                    *count += 1;
                    if (callback != nullptr && (tempCount % INTERVAL) == 0){
                        callback->onProgressUpdate(tempCount);
                    }

                    if (util::isDirectory(children[i].c_str())) {
                        //Log.w(TAG, "Skipping " + subdir.getName() + " subdirectory, " + children[i]);
                        continue;
                    }

                    width = 0;
                    height = 0;
                    isDefined = false;

                    std::string path = util::getFileAsAbsolute(children[i].c_str());
                    if (nitfChecks) {
                        if (PrecisionImageryFactory_isSupported(path.c_str()) == TE_Ok) {
                            do {
                                TAKErr code(TE_Ok);
                                PrecisionImageryPtr image(nullptr, nullptr);
                                code = PrecisionImageryFactory_create(image, path.c_str());
                                TE_CHECKBREAK_CODE(code);
                                ImageInfo info;
                                code = image->getInfo(&info);
                                TE_CHECKBREAK_CODE(code);

                                ul.set(info.upperLeft.latitude, info.upperLeft.longitude);
                                ur.set(info.upperRight.latitude, info.upperRight.longitude);
                                lr.set(info.lowerRight.latitude, info.lowerRight.longitude);
                                ll.set(info.lowerLeft.latitude, info.lowerLeft.longitude);

                                type = info.type;
                                width = info.width;
                                height = info.height;

                                isDefined = true;
                            } while (false);
                        }
                    }
                    if (!isDefined) {
                        GDALDataset *dataset = nullptr;
                        core::Projection *proj = nullptr;
                        try {
                            dataset = (GDALDataset *)GDALOpen(path.c_str(), GA_ReadOnly);
                            if (dataset == nullptr) {
                                //Log.w(TAG, "Unable to open mosaic frame: " + children[i]);
                                continue;
                            }

                            proj = gdal::GdalDatasetProjection::getInstance(dataset);

                            width = dataset->GetRasterXSize();
                            height = dataset->GetRasterYSize();

                            scratch.x = 0;
                            scratch.y = 0;
                            proj->inverse(&scratch, &ul);
                            scratch.x = width;
                            scratch.y = 0;
                            proj->inverse(&scratch, &ur);
                            scratch.x = width;
                            scratch.y = height;
                            proj->inverse(&scratch, &lr);
                            scratch.x = 0;
                            scratch.y = height;
                            proj->inverse(&scratch, &ll);



                            type = genericType;
                        } catch (...) {
                            if (proj)
                                delete proj;
                            if (dataset)
                                GDALClose(dataset);
                            throw;
                        }
                        delete proj;
                        GDALClose(dataset);
                    }
                    double gsd = DatasetDescriptor::computeGSD(width, height, ul, ur, lr, ll);
                    database->insertRow(util::computeRelativePath(relativeTo, children[i].c_str()),
                                       type,
                                       ul,
                                       ur,
                                       lr,
                                       ll,
                                       gsd * (double)(1 << DEFAULT_NUM_LEVELS),
                                       gsd,
                                       width,
                                       height);
                }
            }
            
            
            void MosaicUtils::buildMrsidSubDatabase(ATAKMosaicDatabase *database, const char *subdir,
                                                    const char *relativeTo, int *count,
                                                    BuildMosaicCallback *callback)
            {
                std::vector<std::string> children = util::getDirContents(subdir);
                int width;
                int height;
                core::GeoPoint ul;
                core::GeoPoint ur;
                core::GeoPoint lr;
                core::GeoPoint ll;
                for (size_t i = 0; i < children.size(); i++) {
                    int tempCount = *count;
                    *count += 1;
                    if (callback != nullptr && (tempCount % INTERVAL) == 0){
                        callback->onProgressUpdate(tempCount);
                    }

                    if (util::isDirectory(children[i].c_str())) {
                        //Log.w(TAG, "Skipping " + subdir.getName() + " subdirectory, " + children[i]);
                        continue;
                    }

                    DatasetDescriptor *info = nullptr;
#if 0
                    try {
                        info = MrsidLayerInfoSpi.getLayerInfo(children[i]);
                    } catch (Exception e) {
                        Log.e(TAG, "Error getting mrsid LayerInfo for '" + children[i].getAbsolutePath()
                              + "'");
                        android.util.Log.e(TAG, "error: ", e);
                    }
#endif
                    if (info != nullptr) {
                        auto *image = (ImageDatasetDescriptor *)info;

                        width = static_cast<int>(image->getWidth());
                        height = static_cast<int>(image->getHeight());
                        ul = image->getUpperLeft();
                        ur = image->getUpperRight();
                        lr = image->getLowerRight();
                        ll = image->getLowerLeft();

                        TAK::Engine::Port::String type = info->getImageryTypes()[0];
                        double minGsd = info->getMinResolution(type.get());
                        double maxGsd = info->getMaxResolution(type.get());
                        database->insertRow(util::computeRelativePath(relativeTo, children[i].c_str()),
                                            type.get(),
                                            ul, ur, lr, ll,
                                            minGsd,
                                            maxGsd,
                                            width, height);
                    }
                }
            }
        }
    }
}
